src/engine/enginebufferscalelinear.cpp

```diff
@@ -160,87 +199,34 @@ CSAMPLE* EngineBufferScaleLinear::do_scale(CSAMPLE* buf,
-            SampleUtil::copy(write_buf, &m_bufferInt[iNextSample], readSize);
-            samples_needed -= readSize;
-            write_buf += readSize;
-        }
-
-        // Protection against infinite read loops when (for example) we are
-        // reading from a broken file.
-        bool last_read_failed = false;
-
-        // We need to repeatedly call the RAMAN because the RAMAN does not bend
-        // over backwards to satisfy our request. It assumes you will continue
-        // to call getNextSamples until you receive the number of samples you
-        // wanted.
-        while (samples_needed > 0) {
-            int read_size = m_pReadAheadManager->getNextSamples(
-                    1.0, write_buf, samples_needed);
-            samples_needed -= read_size;
-            write_buf += read_size;
-
-            if (read_size == 0) {
-                if (last_read_failed) {
-                    break;
-                }
-                last_read_failed = true;
-            }
-        }
-
-        // Instead of counting how many samples we got from the internal buffer
-        // and the RAMAN calls, just measure the difference between what we
-        // requested and what we still need.
-        int read_samples = buf_size - samples_needed;
-
-        // Even though this code should not trigger for the special case in
-        // getScaled for when the rate changes directions, the convention in the
-        // rest of this method is that we increment samples_read rather than
-        // assign it.
-        *samples_read += read_samples;
-
-        // Zero the remaining samples if we didn't fill them.
-        SampleUtil::clear(write_buf, samples_needed);
-
-        // update our class members so next time we need to scale it's ok. we do
-        // blow away the fractional sample position here
-        m_bufferIntSize = 0; // force buffer read
-        m_dNextFrame = 0;
-        m_floorSampleOld[0] = buf[read_samples-2];
-        m_floorSampleOld[1] = buf[read_samples-1];
-        return buf;
+    if (rate_diff == 0 && (rate_new == 1.0 || rate_new == -1.0)) {
+        return do_copy(buf, buf_size);
     }
 
     // Simulate the loop to estimate how many frames we need
     double frames = 0;
     // We're calculating frames = 2 samples, so divide remaining buffer by 2;
-    for (int j = 0; j < buf_size / 2; ++j) {
-        frames += (j * 2 * rate_diff / buf_size) + rate_old;
-    }
+    const int bufferSizeFrames = buf_size / 2;
+    const double rate_delta = rate_diff / bufferSizeFrames;
+    // use Gaussian sum formula (n(n+1))/2 for
+    //for (int j = 0; j < bufferSizeFrames; ++j) {
+    //    frames += (j * rate_delta) + rate_old;
+    //}
+    frames = (bufferSizeFrames - 1) * bufferSizeFrames / 2;
+    frames *= rate_delta;
+    frames += rate_old * bufferSizeFrames;
     frames = abs(frames);
 
-    int unscaled_frames_needed = floor(frames);
-
     // If the current position fraction plus the future position fraction
```

Comment:
Comment could use an update -- maybe:
"Intentional integer rounding: increases by one if the fractional part of m_dNextFrame and frames are greater than one"